### ISSUE TITLE
UQ TEST FATAL ERROR, rank 2, in SequenceOfVectorss<V,M>::getPositionValues(): posId is NULL. Exiting...

### ISSUE THREAD

**Issue by dmcdougall** [libqueso/queso#308]
This file produces the error on 4 processors, but not one:

```
#include <queso/GslVector.h>
#include <queso/GslMatrix.h>
#include <queso/UniformVectorRV.h>
#include <queso/StatisticalInverseProblem.h>
#include <queso/ScalarFunction.h>
#include <queso/VectorSet.h>

template<class V, class M>
class Likelihood : public QUESO::BaseScalarFunction<V, M>
{
public:

  Likelihood(const char * prefix, const QUESO::VectorSet<V, M> & domain)
    : QUESO::BaseScalarFunction<V, M>(prefix, domain)
  {
  }

  virtual ~Likelihood()
  {
  }


  virtual double lnValue(const V & domainVector, const V * domainDirection,
      V * gradVector, M * hessianMatrix, V * hessianEffect) const
  {
    return -0.5 * domainVector[0] * domainVector[0];
  }

  virtual double actualValue(const V & domainVector, const V * domainDirection,
      V * gradVector, M * hessianMatrix, V * hessianEffect) const
  {
    return std::exp(this->lnValue(domainVector, domainDirection, gradVector,
          hessianMatrix, hessianEffect));
  }
};

int main(int argc, char ** argv) {
  MPI_Init(&argc, &argv);

  unsigned int dim = 1;

  QUESO::FullEnvironment env(MPI_COMM_WORLD, argv[1], "", NULL);

  QUESO::VectorSpace<QUESO::GslVector, QUESO::GslMatrix> paramSpace(env,
      "param_", dim, NULL);

  double min_val = -1.0;
  double max_val = 1.0;

  QUESO::GslVector paramMins(paramSpace.zeroVector());
  paramMins.cwSet(min_val);
  QUESO::GslVector paramMaxs(paramSpace.zeroVector());
  paramMaxs.cwSet(max_val);

  QUESO::BoxSubset<QUESO::GslVector, QUESO::GslMatrix> paramDomain("param_",
      paramSpace, paramMins, paramMaxs);

  QUESO::UniformVectorRV<QUESO::GslVector, QUESO::GslMatrix> priorRv("prior_",
      paramDomain);

  Likelihood<QUESO::GslVector, QUESO::GslMatrix> lhood("llhd_", paramDomain);

  QUESO::GenericVectorRV<QUESO::GslVector, QUESO::GslMatrix>
    postRv("post_", paramSpace);

  QUESO::StatisticalInverseProblem<QUESO::GslVector, QUESO::GslMatrix>
    ip("", NULL, priorRv, lhood, postRv);

  QUESO::GslVector paramInitials(paramSpace.zeroVector());
  QUESO::GslMatrix proposalCovMatrix(paramSpace.zeroVector());

  for (unsigned int i = 0; i < dim; i++) {
    // paramInitials[i] = 0.0;
    proposalCovMatrix(i, i) = 0.1;
  }

  priorRv.realizer().realization(paramInitials);

  ip.solveWithBayesMetropolisHastings(NULL, paramInitials, &proposalCovMatrix);

  MPI_Finalize();

  return 0;
}
```

What's weird is, if you comment out this line:

```
priorRv.realizer().realization(paramInitials);
```

and uncomment this line

```
// paramInitials[i] = 0.0;
```

then it works fine.  Any ideas?

**Comment by beanSnippet** [comment]
Here is more info about the same or related problem.

The following is included in ChemistryVectorRealizer.C, which is needed for a custom prior. (Most of the code is copied from other built-in VectorRealizer.C files.)
```
const BoxSubset<V,M>* imageBox = dynamic_cast<const BoxSubset<V,M>*>(&m_unifiedImageSet); 

nextValues.cwSetUniform(imageBox->minValues(),imageBox->maxValues());
```
When the second line above is commented out, the code runs on multiple processors. If included, I get this error on 2+ proc, but not 1:
```
UQ TEST FATAL ERROR, rank 1, in SequenceOfVectorss<V,M>::getPositionValues(): posId is
 NULL. Exiting... basic/src/SequenceOfVectors.C, line 207
```

**Comment by beanSnippet** [comment]
Also, for parameter space of size n, code fails on 2+ proc when domainVector[n] appears in JointPdf.C, but not for domainVector[m] with m<n.

**Comment by dmcdougall** [comment]
Good news is I've found where the error happens.  [This](https://github.com/libqueso/queso/blob/dev/src/stats/src/MetropolisHastingsSG.C#L1235) is where the failure happens in the sequence generator.  In particular, [this](https://github.com/libqueso/queso/blob/dev/src/basic/src/SequenceOfVectors.C#L204) is the error that is triggered.

Bad news is the error is triggered on process 1 but not process 0, and I don't know why that's the case.

**Comment by dmcdougall** [comment]
More bad news: I can't recreate this error on `athena`.

**Comment by dmcdougall** [comment]
Alright, I've figured out what's going on here.  The [`unifiedPositionsOfMaximum`](https://github.com/libqueso/queso/blob/dev/src/stats/src/MetropolisHastingsSG.C#L1225) function is supposed to find the indices of the maxima of `workingLogLikelihoodValues` globally over all processors.  After the function exits, the `rawUnifiedMLEpositions` array should contain the samples of the chain at these indices.

The problem is, if on processor zero `workingLogLikelihoodValues` contains the values `[0, 1, 2]` and on processor one it contains `[10, 11, 12]`, then after the `unifiedPositionsOfMaximum` function exits the `rawUnifiedMLEpositions` array is of length one on processor one contains a single pointer to a sample.  This is because there is a unique maximum on processor one: it is 12.  On processor zero it is also of length one, but the element is `NULL`.  This is because there is no maximum on processor zero since all the elements on processor one are bigger.  Therefore there is no corresponding sample in the chain that corresponds to this maximum, so a null pointer is returned.  This is probably the right thing to do, [the issue comes later when queso assumes the opposite](https://github.com/libqueso/queso/blob/dev/src/stats/src/MetropolisHastingsSG.C#L1235).

[I have implemented a test case that hits this](https://github.com/libqueso/queso/pull/314).  It's basically a copy-paste of the relevant lines in `MetropolisHastingsSG.C`.  Some thought is needed to decide the best way to 'fix' this.  I use quotes around 'fix' because in my opinion the function does what it is supposed to do, although the documentation for this function is threadbare so we could argue this is a documentation bug.  Perhaps the best way to go about 'fixing' this is to surround the [`getPositionValues`](https://github.com/libqueso/queso/blob/dev/src/stats/src/MetropolisHastingsSG.C#L1235) call in a `try`/`catch` block and add a comment summarising what I have explained above.

Input very much appreciated, though, as I'm sure there are other ways to tackle this problem.

**Comment by dmcdougall** [comment]
Furthermore, the `try`/`catch` block method only addresses this particular call, with no regard to use of this array later on in the subroutine.

**Comment by pbauman** [comment]
You sure the intent wasn't for the maximum that is found to be sent to all processors? i.e. MPI_Allreduce instead of MPI_Reduce?

**Comment by dmcdougall** [comment]
I'm not sure of anything at this point.

You make an excellent observation.  Although it would mean, at present, this function would give processor zero a pointer to an element that exists on processor one.  This violates the assumption of no global address space in MPI.

**Comment by pbauman** [comment]
> Although it would mean, at present, this function would give processor zero a pointer to an element that exists on processor one.

Sorry, I'm not looking at the code, but you were talking about values (numbers) above. So, an Allreduce (computing the max in this case) will find the max and then send the value to every processor. Everybody had their own copy of the same value then. Make sense?

**Comment by dmcdougall** [comment]
>Sorry, I'm not looking at the code, but you were talking about values (numbers) above. So, an Allreduce (computing the max in this case) will find the max and then send the value to every processor. Everybody had their own copy of the same value then. Make sense?

Sorry, there are two lines of discussion going on here.

That makes sense, yes.  That would also be a reasonable solution.  Bear in mind that this solution requires a change to the function calling signature.

I was proposing to modify behaviour while keeping the current calling signature.  This way it's 'backwards compatible' because the documentation is so poor that we can make an argument as to what the intended behaviour should have been.

Also, as an aside, it's surprising this wasn't hit earlier considering that having the same sample on more than one processor is a probability zero event (assuming the seed is set to something sensible).  Furthermore, the above example works on queso v0.50.1 but not on v0.51.0, implying there was a regression.  I haven't looked into finding the offended commit yet.

**Comment by dmcdougall** [comment]
>Furthermore, the above example works on queso v0.50.1 but not on v0.51.0, implying there was a regression. I haven't looked into finding the offended commit yet.

The test I wrote in #314, however, still fails in v0.50.1.

**Comment by dmcdougall** [comment]
>Furthermore, the above example works on queso v0.50.1 but not on v0.51.0, implying there was a regression. I haven't looked into finding the offended commit yet.

`git bisect` tells me that 9ad7de3e3e948fd28a5e48ce6e805fb2ee49e724 is the offending commit.

**Comment by dmcdougall** [comment]
>>Furthermore, the above example works on queso v0.50.1 but not on v0.51.0, implying there was a regression. I haven't looked into finding the offended commit yet.
>
> `git bisect tells` me that 9ad7de3 is the offending commit.

More specifically, [this](https://github.com/libqueso/queso/commit/9ad7de3e3e948fd28a5e48ce6e805fb2ee49e724#diff-8727a3ccd3f94bca308acdf678e6e202L261) is the offending change within the commit that allows the branch [here](https://github.com/libqueso/queso/blob/dev/src/stats/src/MetropolisHastingsSG.C#L1223) to get hit.

**Comment by dmcdougall** [comment]
This should be fixed by #314.  I'll write a little summary of the changes to `unifiedPositionsOfMaximum` shortly.

**Comment by dmcdougall** [comment]
Fixed in #314.
